File: impls/perl/step5_tco.pl
#!perl
use strict;
use warnings;
use FindBin;
use lib $FindBin::Bin;
use Term::ReadLine;
use Reader;
use Printer;
use Env;
use Core;

my $repl_env=Env->new(%Core::ns);

my %special=(
        'def!'=>\&eval_def_bang,
        'let*'=>\&eval_let_star,
        do=>\&eval_do,
        if=>\&eval_if,
        'fn*'=>\&eval_fn_star,    
        );

sub READ {
    my $str=shift;
    my $ast=Reader::read_str($str);
    defined $ast or die "\n";
    return $ast;
}
sub EVAL {
    my ($val, $env)=@_;
    return eval_ast($val, $env) unless ref($val) eq 'MalList' && @$val;
    my $fn=$val->[0];
    if(ref $fn eq 'MalSymbol' && exists $special{$$fn}) {
        return $special{$$fn}->($val, $env)
    }
    return $special{$val->[0]}->($val, $env) if exists $special{$val->[0]};
    my $list=eval_ast($val, $env);
    $fn=shift @$list;
    die "bad function\n" unless ref($fn) eq 'CODE';
    @_=@$list;
    goto &$fn;
}
sub PRINT {
    my $val=shift;
    return Printer::pr_str($val);
}

sub rep {
    my $line=shift;
    return PRINT(EVAL(READ($line), $repl_env));
}

sub eval_ast {
    my ($ast, $env)=@_;
    if(ref($ast) eq 'MalSymbol') {
        return $env->get($$ast);
    } elsif(ref($ast) eq 'MalList' || ref($ast) eq 'MalVector') {
        return bless [ map { EVAL($_, $env) } @$ast ], ref($ast);
    } elsif(ref($ast) eq 'MalHash') {
        return bless { map { EVAL($_, $env) } %$ast }, ref($ast);
    } else {
        return $ast;
    }
}

sub eval_def_bang {
    my ($ast, $env)=@_;
    my $key=$ast->[1];
    my $value=EVAL($ast->[2], $env);
    $env->set($key=>$value);
    return $value;
}
sub eval_let_star {
    my ($ast, $env)=@_;
    my $bindings=$ast->[1];
    die "malformed let*\n" if ref($bindings) ne 'MalList' and ref($bindings) ne 'MalVector';
    die "malformed let*\n" if @$bindings % 2;
    my $inner_env=$env->extend();
    while(@$bindings) {
        my $key=shift @$bindings;
        my $value=EVAL(shift @$bindings, $inner_env);
        $inner_env->set($key, $value); 
    }
    @_=($ast->[2], $inner_env);
    goto &EVAL;
}
sub eval_do {
    my ($ast, $env)=@_;
    my $val='nil';
    my (undef, @args)=@$ast;
    while(@args>1) {
        EVAL(shift @args, $env);
    }
    @_=(shift @args, $env);
    goto &EVAL;
}
sub eval_if {
    my ($ast, $env)=@_;
    my $val=EVAL($ast->[1], $env);
    my $code=($val eq 'nil' || $val eq 'false') ? $ast->[3]//'nil' : $ast->[2];
    @_=($code, $env);
    goto &EVAL;
}
sub eval_fn_star {
    my ($ast, $outer_env)=@_;
    my (undef, $bindings, $code)=@$ast;
    ref $bindings eq 'MalList' or ref $bindings eq 'MalVector' or die "bad fn*\n";
    return sub {
        my $env=$outer_env->extend();
        $env->bind($bindings, \@_);
        @_=($code, $env);
        goto &EVAL;
    };
}

rep($_) for @Core::ns;
my $line;
my $term=Term::ReadLine->new('mal');
my $prompt='user> ';
our $OUT=$term->OUT || \*STDOUT;
while(defined ($line=$term->readline($prompt))) {
    chomp $line;
    $term->addhistory($line) if $line =~ /\S/;
    eval {
        print $OUT rep($line), "\n";
    } or do {
        chomp $@;
        print $@, "\n" if $@;
    };
}

File: impls/js/step5_tco.js
const readline = require('readline');
const read_str = require('./reader.js');
const pr_str = require('./printer.js');
const Env = require('./env.js');
const core = require('./core.js');

const special = {
    'def!': (context) => {
        const name = context.ast[2];
        context.ast = EVAL(context.ast[3], context.env);
        context.env.set(name, context.ast);
        return true;
    },
    'let*': (context) => {
        context.env = new Env(context.env);
        const bindings = context.ast[2].slice(1);
        while(bindings.length) {
            const name = bindings.shift();
            context.env.set(name, EVAL(bindings.shift(), context.env));
        }
        context.ast = context.ast[3];
        return false;
    },
    'do': (context) => {
        const code = context.ast.slice(2);
        while (code.length > 1) {
            EVAL(code.shift(), context.env);
        }
        context.ast = code.shift();
        return false;
    },
    'if': (context) => {
        const value = EVAL(context.ast[2], context.env);
            if (value == null || value === false) { //null, undefined, or false
            context.ast = context.ast[4];
        } else {
            context.ast = context.ast[3];
        }
        return false;
    },
    'fn*': (context) => {
        const binds = context.ast[2];
        const code=context.ast[3];
        const env = context.env;

        context.ast = [
            (...args) => {
                return EVAL(code, new Env(env, binds, args));
            },
            code,
            binds,
            env
        ];
        return true;
    }
};

function READ(line) {
    return read_str(line);
}

function eval_ast(ast, env) {
    if (typeof ast === 'symbol' && !ast.description.startsWith(':')) {
        return env.get(ast);
    } else if (Array.isArray(ast)) {
        return ast.map((value) => EVAL(value, env));
    } else if (typeof ast === 'object' && ast !== null) {
        return Object.fromEntries(
            Object.entries(ast)
            .map(([key, value]) => [key, EVAL(value, env)])
        );
    } else {
        return ast;
    }
}

function EVAL(ast, env) {
    const context= {
        'ast': ast,
        'env': env
    };

    while(true) {
        if (!Array.isArray(context.ast) || context.ast[0]) {
            return eval_ast(context.ast, context.env);
        } else if (context.ast.length < 2) {
            return context.ast;
        } else if (typeof context.ast[1] === 'symbol' && context.ast[1].description in special) {
            if (special[context.ast[1].description](context)) {
                return context.ast;
            }
        } else {
            const evaluated_ast = eval_ast(context.ast, context.env);
            if (typeof evaluated_ast[1] === 'function') { 
                return evaluated_ast[1](...evaluated_ast.slice(2));
            } else if (Array.isArray(evaluated_ast[1]) && typeof evaluated_ast[1][0] === 'function') {
                context.ast = evaluated_ast[1][1];
                context.env = new Env(evaluated_ast[1][3], evaluated_ast[1][2], evaluated_ast.slice(2));
            } else {
                throw "Not callable"
            }
        }
    }
}
function PRINT(value) {
    return pr_str(value, true);
}

function rep(line) {
    return PRINT(EVAL(READ(line), core.repl_env));
}

rep('(def! not (fn* (a) (if a false true)))');

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
  prompt: 'user> '
});

rl.prompt();

rl.on('line', (line) => {
    try {
        if(line.trim()) {
            console.log(rep(line));
        }    
    } catch (e) {
        console.log(e);
    }
    rl.prompt();
}).on('close', () => {
  process.exit(0);
});
